File: js/factorial/factorial.py
level = 0

def fac(n):
    print "{ action: 'push', line: 0, text:'fac(n=%d)' }," % (n);
    print "{ action: 'step' }," 
    print "{ action: 'curline', line: 1,    text:'   if %d <= 1'}," % (n)
    print "{ action: 'step' }," 
    if n <= 1:
        print "{ action: 'curline', line: 2,text: '      return 1'}," 
        print "{ action: 'step' }," 
        print "{ action: 'pop' }," 
        print "{ action: 'step' }," 
        return  1
    print "{ action: 'curline', line: 3, text:'   f = fac(%d-1)' }," % (n)
    print "{ action: 'step' }," 
    f = fac(n-1)
    print "{ action: 'line', line: 3,    text:'   f = %d' }," % (f)
    print "{ action: 'step' }," 
    print "{ action: 'curline', line: 4, text:'   f = %d * %d' }," % (f, n)
    f = f * n;
    print "{ action: 'step' }," 
    print "{ action: 'curline',  line: 5,text:'   return %d'}," % f;
    print "{ action: 'step' }," 
    print "{ action: 'pop' }," 
    return f


import random; 

if __name__ == "__main__":

    for n in [0, 1, 3, 5, 7]: 
       print "AboutAlgo[%d]  = { n: %d };"  % (n,n);
       print "AboutAlgo[%d].name = '%d Factorial';" % (n, n);

    for n in [0, 1, 3, 5, 7]: 
       print "AboutAlgo[%d].steps = [" % (n); 
       fac(n); 
       print "];"; 
       print "\n"; 

File: factorial.js
var i;

var TheTrace = undefined;
var TheLineCounter = undefined;
var LineCounter = undefined;
var TheStackTrace = undefined;

var TheLevelCounter = undefined;
var LevelCounter = undefined;

var TheActionList = new ActionList();

var AlgoIndex = undefined;
var TheAlgo = undefined;

function doRun() {
	var isStep = TheActionList.doNextStep();
	while (isStep != undefined) {
		isStep = TheActionList.doNextStep();
	}
	return isStep;
}


function doStep(action_obj) {
	// TheStackTrace.getTopCodeBlock().Wrapper.css("background-color")
	if (action_obj.action == "step") {
		return;
	}
	if (action_obj.action == "pop") {
		// TheRangeCharts[TheStackTrace.top].Wrapper.remove();
		TheStackTrace.pop();
		TheTrace.setCurLine(undefined);
		var top = TheStackTrace.getTopCodeBlock();
		if (top) {
			top.setCurLine(undefined);
		}
		return;
	}
	if (action_obj.action == "push") {
		var level = TheStackTrace.pushCode(TheAlgo.code);
		if (LevelCounter.length <= level) {
			LevelCounter[level] = 1;
			TheLevelCounter.appendLine("1");
		} else {
			LevelCounter[level]++;
			TheLevelCounter.setCurLine(level, "" + LevelCounter[level]);
		}
	}
	if (action_obj.action == "curline" || action_obj.action == "push") {
		TheTrace.setCurLine(action_obj.line);
		LineCounter[action_obj.line] = LineCounter[action_obj.line] + 1;
		TheLineCounter.setCurLine(action_obj.line, "" + LineCounter[action_obj.line]);
		TheStackTrace.getTopCodeBlock().setCurLine(action_obj.line, action_obj.text);
	    return;
	}
	if (action_obj.action == "line") {
		TheTrace.setCurLine(action_obj.line);
		TheStackTrace.getTopCodeBlock().setCurLine(action_obj.line, action_obj.text);
	    return;
	}
	assert(false, "should never get here");
	return;
}

function resetButtons(actionList) {
	var nxt = $("#next");
	var run = $("#run");
	var stop = $("#stop");
	var reset = $("#reset");

	var opacity = (actionList.actionAllowed(actionList.doMultiSteps)) ? 1.0 : 0.5;
	nxt.css("opacity", opacity);
	opacity = (actionList.actionAllowed(actionList.doRun)) ? 1.0 : 0.5;
	run.css("opacity", opacity);
	opacity = (actionList.actionAllowed(actionList.doStop)) ? 1.0 : 0.5;
	stop.css("opacity", opacity);
	opacity = (actionList.actionAllowed(actionList.reset)) ? 1.0 : 0.5;
	reset.css("opacity", opacity);
}

function animate(fct) {
	console.log('anaimate');
	var suc = fct();
}


function chooseProblem(algo, index) {
	if (!TheActionList.actionAllowed(TheActionList.reset)) {
		return undefined;
	}

	if (TheTrace) {
		TheTrace.Wrapper.remove();
	}
	TheTrace = new CodeBlock($("#TheCode"), algo.code);

	if (TheStackTrace) {
		TheStackTrace.Wrapper.remove();
	}
	TheStackTrace = new CodeStack($("#TheStackTrace"), algo.code, $("#ColorPalette"));

	if (TheLineCounter) {
		TheLineCounter.Wrapper.remove();
	}
	TheLineCounter = new CodeBlock($("#TheLineCounter"), algo.code.length, "0");
	LineCounter = new Array();
	for ( i = 0; i < algo.code.length; i++) {
		LineCounter.push(0);
	}

	if (TheLevelCounter) {
		TheLevelCounter.Wrapper.remove();
	}
	LevelCounter = new Array();
	LevelCounter.push(1);
	TheLevelCounter = new CodeBlock($("#TheLevelCounter"), 1, "1");

	$("#algo").html(algo.name);
	//$("#log").html("LOG");

	if (TheAlgo == algo && AlgoIndex == index) {
		TheActionList.reset();
	} else {
		TheActionList = new ActionList();
		for (var i = 0, l = algo[index].steps.length; i < l; i++) {
			action = algo[index].steps[i];
			TheActionList.pushStep(doStep, (action.action == "step") ? 200 : 0, [action]);
		}
	}

	$("#TheAlgoName").html(algo[index].name);
	$('#TheStackTrace').width($("#TheCode").width());
	AlgoIndex = index;
	TheAlgo = algo;
	TheActionList.doMultiSteps();
	resetButtons(TheActionList);
	TheActionList.stateChange = resetButtons;
	return true;
}

function onLoad() {
	params = getUrlParams();
	n = parseInt(params.n, 10);
	if (isNaN(n)) { n = 3; }
	if (n < 0) { n = 0; }
	if (n > 7) { n = 7 }
	if (AboutAlgo[n] === undefined) {
		n = 3;
	}
	chooseProblem(AboutAlgo, n);
}

